### PR TITLE
gh-98703: Fix asyncio proactor_events calling _call_connection_lost multiple times

### PR DESCRIPTION
**Pull request by Fidget-Spinner**
* Issue: gh-98703

### CODE REVIEW

**Comment by Fidget-Spinner** [comment]
@kumaraditya303 I don't know why but this started showing on our tests after https://github.com/python/cpython/pull/98572. I can't tell if that PR introduced a bug or it just exposed something wrong with our implementation.

IMO, it seems it's something wrong with our implementation.

**Comment by kumaraditya303** [comment]
`asyncio` always has surprises hence it is my favorite :)

Seriously though this has uncovered another race in our implementation. The code does not guard against concurrent closing of the stream which leads to this failure. The `connection_lost` callback should be called only once and here it is being called more than once.

Here's a better fix:
```diff
diff --git a/Lib/asyncio/proactor_events.py b/Lib/asyncio/proactor_events.py
index 2685a3376c..c6aab408fc 100644
--- a/Lib/asyncio/proactor_events.py
+++ b/Lib/asyncio/proactor_events.py
@@ -152,6 +152,8 @@ def _force_close(self, exc):
         self._loop.call_soon(self._call_connection_lost, exc)
 
     def _call_connection_lost(self, exc):
+        if self._called_connection_lost:
+            return
         try:
             self._protocol.connection_lost(exc)
         finally:

```

Feel free to update this PR with the patch above, I have verified this.

File: Misc/NEWS.d/next/Library/2022-10-26-07-51-55.gh-issue-98703.0hW773.rst [hunk 1]
```diff
@@ -0,0 +1 @@
+Fix asyncio ``drain`` closing a non-existent socket multiple times.
```
Comment (kumaraditya303) [inline Misc/NEWS.d/next/Library/2022-10-26-07-51-55.gh-issue-98703.0hW773.rst]:
```suggestion
Fix :meth:`asyncio.StreamWriter.drain` to call ``protocol.connection_lost`` callback only once on Windows. 
```

**Comment by Fidget-Spinner** [comment]
> Feel free to update this PR with the patch above, I have verified this.

Awesome! Thank you for the fix. I was trying to find something more elegant but I'm a noob at asyncio internals :). Your fix is way nicer.

File: Lib/asyncio/proactor_events.py [hunk 2]
```diff
@@ -161,7 +163,6 @@ def _call_connection_lost(self, exc):
             # cure it, but maybe using DisconnectEx() would be better.
             if hasattr(self._sock, 'shutdown') and self._sock.fileno() != -1:
                 self._sock.shutdown(socket.SHUT_RDWR)
-            self._sock.close()
```
Comment (kumaraditya303) [inline Lib/asyncio/proactor_events.py]: This doesn't need to be removed else it'll leak the socket.

File: Lib/asyncio/proactor_events.py [hunk 3]
```diff
@@ -161,7 +163,6 @@ def _call_connection_lost(self, exc):
             # cure it, but maybe using DisconnectEx() would be better.
             if hasattr(self._sock, 'shutdown') and self._sock.fileno() != -1:
                 self._sock.shutdown(socket.SHUT_RDWR)
```
Comment (kumaraditya303) [inline Lib/asyncio/proactor_events.py]:
```suggestion
                self._sock.shutdown(socket.SHUT_RDWR)
                self._sock.close()
```

File: Lib/asyncio/proactor_events.py (hunk 2 above)
Comment (Fidget-Spinner) [reply Lib/asyncio/proactor_events.py]: Yep that was accidental. Thanks

**Review by kumaraditya303 — APPROVED** [review]
LGTM thanks for the PR!

**Comment by kumaraditya303** [comment]
I created https://github.com/python/cpython/pull/98730 to add tests for this.

**Review by gvanrossum — APPROVED** [review]
Ironic.

**Pull request merged**